SCI32: Fix GK1 honfour keycard lockup

sluicebox opened the pull request:
Fixes lockup in original game, bug #10767

m-kiewitz commented:
Hi,


I just can't login from work. The patch looks fine, documentation also looks fine, but PATCH_ADDTOOFFSET(+x) should be used instead of copying over all the byte code into the patch data.


Kind Regards

  M. Kiewitz


> David Turner <notifications@github.com> hat am 24. Oktober 2018 um 12:41 geschrieben:
> 
> 
>     @digitall https://github.com/digitall requested your review on: #1363 https://github.com/scummvm/scummvm/pull/1363 SCI32: Fix GK1 honfour keycard lockup.
> 
>     —
>     You are receiving this because your review was requested.
>     Reply to this email directly, view it on GitHub https://github.com/scummvm/scummvm/pull/1363#event-1923021255 , or mute the thread https://github.com/notifications/unsubscribe-auth/AFNjcXfcxWbzwrlfrM8ySxiVtUZYM-wUks5uoEPzgaJpZM4X24hV .
>

digitall commented:
@m-kiewitz : Now that is a feature I haven't seen used before! :)

Thanks for the review. @sluicebox: OK to rework using PATCH_ADDTOOFFSET(+x) ?

m-kiewitz commented:
PATCH_ADDTOOFFSET(+x) simply adds a number to the patch offset, which lets you basically skip bytes.
The feature can also be useful when an offset/id is used within the script code, which isn't the same for all versions. There is also a feature, that let's you copy over bytes from the script code and allows one to copy them to a different offset. And there is also a feature, that gets a uint16 from the original script code and adds a specified number to it. That let's one to adjust offsets for example.

m-kiewitz commented:
@sluicebox btw. interested in officially joining our team?

bluegr commented:
There are four things I wanted to point out here:
- Great work, as usual!
- If this bug has already been fixed by an official Sierra patch, is there any reason to spend time fixing it again?
- PATCH_ADDTOOFFSET helps in a lot of ways, as @m-kiewitz mentioned
- Same question from me, @sluicebox would you like to officially join our team? :)

bluegr commented on engines/sci/engine/script_patches.cpp:
```diff
@@ -1431,6 +1431,34 @@ static const uint16 gk1CazanouxPathfindingPatch[] = {
+	0x7c,                           // pushSelf
+	0x81, 0x5b,                     // lag 5b
+	0x4a, SIG_MAGICDWORD,           // send e [ gkMessager:say ... self ]
+	SIG_UINT16(0x000e),
```
Nitpicking: the above two lines should be merged in one, since they form the same command, like so:

```suggestion
	0x4a, SIG_MAGICDWORD, SIG_UINT16(0x000e), // send e [ gkMessager:say ... self ]
```

sluicebox commented:
Wow, a lot going on on this one!

I would be honored to join!! Keeping Sierra games running is the only legitimate use of computers anymore.

We're on the same page regarding bugs Sierra fixed. I'm mentioning the floppy patches to avoid confusion in case some power developer is testing with a floppy version that they've applied an official patch to. In my experience it's unlikely that a normal user has gone and applied external Sierra patches, assuming they even knew that was possible. It's more realistic to just say "fixed in CD version" in which case I consider all fixable takes-you-out-of-the-game bugs in floppy versions fair game, fixed in CD or not, like lock-ups, missing points (so frustrating), and error messages. That said, I can read a room, and I'll see if I can rustle up an unpatched one next =)

I like PATCH_ADDTOOFFSET but I'm interested to see how this one could be rewritten use that. I don't think any bytes remain in their original position plus some of the re-ordered instructions are optimized (like pushi 0005 => pushi 05). I'm only looking at existing patches for examples though so I'm sure I'm missing some of the script patcher's power. (btw thanks for making such a cool script patcher!)

m-kiewitz commented:
@sluicebox 
Then I guess welcome on board?
I guess you will need to contact @sev- for IRC credentials (please get yourself a cloak as well, sev can get you into contact, that way your IP address won't be visible to other people) plus ability to commit for yourself and so on.

Maybe I'm a bit confused, but it seems this script patch changes the first byte/opcode and the last opcode (byte + uint16) and everything else stays the same? So one could do a PATCH_ADDTOOFFSET(+5) to skip all of those unmodified bytes.

Fully agreed on that external patching. Most users do not know and my opinion is that they shouldn't need to know. The game should simply run properly with as many game bugs fixed as possible and no user actions required.

Fun fact about that patcher code - it's based on my x86 VM patcher, which was capable of fixing quite a few run-time errors in old DOS programs on-the-fly as well, that happened to occur when the computer was too fast for the run-time library. For example "Run Time Error 200 (Divide by Zero)". Really useful especially for cases where the .EXE is compressed and can't be uncompressed and thus file patching is not possible.

The code was re-used and adjusted accordingly after we originally put in extra C++ code for every single fix of script bugs, which got really messy pretty fast. This here also means that people can extract the patches and then apply them manually to their games / simply dump the patched script, so that the bug is fixed even when using the original interpreter. Plus one patch can cover 5 game versions all at the same time, which a file patch can't unless the script is exactly the same for all versions (major issue with some fan patches, some fan patches even break certain other versions of games).

For script patching sometimes there aren't enough bytes available to add additional script code even when using all sorts of funky script optimization tricks.
I thought about ways to add additional script code space. Maybe one could add a few bytes at the end of the script, but it gets a bit complicated, because script sizes are not the same quite often for different game versions, so jumps would have to get adjusted accordingly and of course I would prefer some kind of generic solution, that doesn't require manual adjustments. The signatures + patch data right now are also not fit for that job. There would also be a way adding another segment, but that's also problematic because of the way SCI handles segments.

Script patches are always meant to be fully optional. What this means is that the game should work fine even after applying a patch, running the game, saving, then removing the patch and restoring the saved game. This is why adding an additional code segment wouldn't work out. We would break saved games.

sluicebox commented:
I see how this patch looks that way but it's more complicated. It inserts 13 new bytes to award the points, which is a tall order that looked hopeless, but I got lucky in how the script is structured:

```
(if (isFlag 27)
	(GKEgo getPoints: -999 2) ; the new line
	(setFlag 26)
	(GkMessager say: 17 16 56 0 self) ; "Really? So that's what Hartridge used..."
else
	(GkMessager say: 17 16 55 0 self) ; "Really? So this is the whole pattern..."
)
```

Since both code paths end in a similar call with only one different parameter I moved the initial common pushes for setting up the say call in front of the flag test to buy a lot of bytes. In the process I optimized one of those pushes to save another byte. Then I optimized the bnt to save another byte and changed the destination to the push 55 after the patch and complete the say call. After the new instructions award the points I push 56 for that code path's cond and also jump into the middle of that second say (after the push 55) and let it complete the call.

It's clearer if you can see the instructions after the patch, but I know it's still complicated.

sluicebox commented:
That's great that your patching technique goes that far back. Patching Sierra instructions is a lot more fun for me than x86 was.
